Add get_grid method.

charrea6 opened the pull request:
get_grid returns programs split into channels, with additional metadata added to the programs by registered callbacks.

The intention is that this is used to retrieve the grid for display with the programs also including information on whether they have been scheduled for recording/are a favorite.

charrea6 committed:
Add get_grid function to guide.
get_grid returns programs split into channels, with additional metadata added to the programs by registered callbacks.

The intention is that this is used to retrieve the grid for display with the programs also including information on whether they have been scheduled for recording/are a favorite.

jtackaberry commented on test/grid.py:
```diff
@@ -0,0 +1,39 @@
+@kaa.coroutine()
+def remote():
+    g = kaa.epg.connect(('localhost', 10000))
+    yield g.signals.subset('connected').any()
```
This is more direct:

``` python
yield kaa.inprogress(g.signals['connected'])
```

charrea6 replied:
Nice I tried a couple of other attempts but missed this. I'll update the code,

jtackaberry commented:
I'm struggling to see the value of the grid callbacks.  It sounds like it's just a way for the caller to avoid iterating over the returned lists to do custom stuff.  Can you help me understand the use-case?

Why use Meta instead of a simple dict?  In addition to being simpler (IMO) it's also cheaper to create `{}` than `object()`

Moving the `convert()` function out of `search()` is sensible but should be a separate commit, since it's unrelated to this pull request.  Also it should be done for guide.py.  I think there is an opportunity to use `kaa.dateutils.to_timestamp()` which was recently added to kaa.base.

Lastly, we have a lot of ugly code duplication between guide.py and rpc.py.  I appreciate you're just following the poor precedent set by `search()` but adding another function with the a lot of duplicated code just underlines the fact that this needs to be refactored.

It'd be great if that could be done before adding this `get_grid()` feature, but if you prefer to do it some other time, that's ok, but in that case I'd not bother duplicating the docstring in guide.py and instead put a FIXME.

Thanks!

charrea6 commented:
The point of the callbacks is to allow an EPG client to retrieve the grid data from a server over RPC, where the server also contains info on the favorite/schedule status. This reduces the number of RPC calls that have to be made as you don't have to do a separate call to get info on the schedule or favorites, ie one point of information, the record server.

As for the Meta rather than the dict, well personal choice I suppose and I was considering what was already in Freevo 1, ie its an attribute access to retrieve the info. I prefer the object attribute approach, 1 it's less typing to access a value and 2 it just feels more pythonic. YMMV

There was a reason for moving the convert() in this commit it was going to be duplicated code, I'd missed that it was also in guide.py. I'll take a look at the to_timestamp and see if it does the same thing....

I'll take another look at rpc.py and see what can be refactored....

jtackaberry commented:
Ah right, makes sense, thanks.  I didn't pay enough attention to your example where it makes it clear the callback registration is intended to be done server side.

I would prefer we use a dict here instead of a Meta instance.  And let's call it userdata, as it's not really metadata (i.e. it's not data about data, it's just additional data related to the Program).  This feels more conventional to me.

I'd then formalize the attribute by initializing it to None in the Program class.

jtackaberry commented on src/guide.py:
```diff
@@ -226,6 +227,87 @@ def combine_attrs(row):
+        for row in programs:
+            # Populate the meta object
+            meta = Meta()
+            for callback in self._grid_callbacks:
```
Should not bother creating a Meta object (or dict, as the case may be) unless _grid_callbacks is not empty.

charrea6 commented:
I'll change the meta object to a dict as although it's longer to type known attributes it's not so easy to do the 'default setting if not available'.

As for the name I don't like userdata, far too vague were as I feel meta is correct as it is data about data, ie in this case its data about the program object, if it is schedule to record, whether is a favorite etc. Its not additional data on the program which should be additional attributes.

jtackaberry commented:
The term "userdata" is commonly used in libraries and has a clear meaning (if not formal, definitely de facto): it's data that's not generated from within the library, rather it's opaque data provided by the caller elsewhere and merely passed along by the library.

I don't think stuff like "is this program scheduled to be recorded" constitutes metadata.   If you define "data" in this context to be the program and "metadata" to be "data about the program" then even the program title and description would qualify as metadata under that definition.  I think whether the program is a favourite is in the same category as genre and rating, which are attributes of the Program object.  To me, metadata in this context would be stuff like the source of the program data (schedulesdirect.org, say), or when it was last updated from the server, etc.

So IMO "userdata" is much less vague than "metadata" because at least has _some_ descriptive value.

charrea6 commented:
Another reason I don't like userdata is that it suggests its being added by the caller somehow or is some sort of callback data (originating from the user of the API), which it isn't. 
The server is providing it from another part of the server (ignoring the local case here I know), therefore its not user data, it may however be additional data. So how about a compromise additional_info rather that userdata or meta.

jtackaberry commented:
Maybe I'm missing something important here.  I thought the actual data _is_ being originated by the user of the API?  It seems to me that the client/server distinction is a red herring.  The recordserver is a user of the kaa.epg API, extending the data passed back to the client via the grid callback.

So from kaa.epg's perspective, this _is_ userdata.  I think your objection is that from the _client_ side, this doesn't smell at all like userdata because the server provided that data, and it doesn't necessarily know or appreciate the fact that the server actually extended kaa.epg to provide that extra info.

Does that seem like a fair statement?

I usually try to avoid attribute names with underscores if possible.  So maybe `extrainfo`?  Another option we might consider is whether it makes sense to just have `Client.get_grid()` merge the extrainfo dict straight into the program object's `__dict__`.  If the goal is to be able to extend the Program object in a way that the client shouldn't have to know or care about, this might be acceptable (although it's not without its downside).  What do you think?

I don't want to wear you down on this stuff, because I do appreciate your contribution.  I've voiced my opinion and provided a couple other options, so go ahead and do what you think is right. :)

charrea6 commented:
Yes and Yes, I beginning to think about this not best way to do things but am struggling to see a better solution.
I've got a solution currently just for Freevo 1 where this call is in the recordserver RPC context but this is ugly and doesn't give properly created EPG Program objects back, channel for one is probably duplicated rather than a ref to the client context channel.

extrainfo sounds fine but, and here's a doh moment, what about gridinfo?
I think keeping it separate is best as it is obvious that these extra bits are only available when you call get_grid.

Don't worry about wearing me down, I'm used to working in a place where everything is code reviewed (down to single character changes) so this isn't new to me, plus sometimes it the best way to get a good API.

jtackaberry commented:
`gridinfo` is fine with me if you think this approach of having callbacks attach additional data to Program objects is going to stay unique to `get_grid()`?  Do you think there might be other times where Program objects would be augmented with external data?

Maybe it makes sense to make this completely generic, so that _any_ time a Program object is returned by the API, it invokes user specified callbacks to attach new data?  In that case it could be reasonable to merge the extrainfo dict into the program object's `__dict__` because those attributes would be available in all cases.

charrea6 commented:
That's an interesting point as it would conceivable make sense for `search()` to return the same sort of information as `get_grid()`, ie transient information such as scheduled/favorite status. You'd want to know that information in the UI so you could modify the behaviour appropriately.

Perhaps the best thing to do therefore is to pass this extra information to the `Program` class on creation so that it can either add it to the object as attributes or as a single attribute containing a dictionary. For the standard `Program` class it gets added to the object dictionary.

Ok, I'll go with that and expand the behaviour to search as well.

charrea6 committed:
Remove get_grid and change search to call the callbacks instead.
This way all queries of the database will have access to the additional information. Also remove the meta object and instead apply the returned properties to the Program object as attributes.

 Modifiy the grid test program to separate the programs itself and then check the test value as before.

jtackaberry commented on src/guide.py:
```diff
@@ -211,20 +201,55 @@ def convert(dt):
-            results.append(cls(channel, row))
-        yield results
+        for i,row in enumerate(query_data):
+            channel = self._channels_by_db_id[row['parent_id']]
```
You seem to have removed some of the verifications done on channel and row['parent_id'].  Did you determine that those checks are unnecessary?

jtackaberry commented on src/guide.py:
```diff
@@ -302,3 +327,9 @@ def get_genres(self, associated=None, prefix=None):
         return self._num_programs
+
+def to_timestamp(dt):
+    'Converts a time to a unix timestamp (seconds since epoch UTC)'
```
Different docstring formatting.  As a matter of style, kaa uses `"""` everywhere (with `"""` always being on dedicated lines).  (Ok, this looks like it was part of the original inline function.  My bad :))

jtackaberry commented:
I prefer this now, much more general purpose.

Looking at the register/unregister callback stuff, I now wonder if we shouldn't instead us a Signal.  This is just what signals are: something happens and we want to give the user a way to attach/detach callbacks to do stuff when that thing happens.  In this case, Guide would subclass kaa.Object and define a `__kaasignals__` attribute with something like a `new-program` signal.  (You can see sockets.py in kaa.base for a good example.)

By reusing the Signal machinery, this should simplify the patch even more, and make it more kaa-like.

jtackaberry commented on src/guide.py:
```diff
@@ -211,20 +201,55 @@ def convert(dt):
+                extra_info = None
+            else:
+                extra_info = extra_data[i]
+            results.append(cls(channel, row, extra_info))
```
extra_info vs extra_data is somewhat confusing to read.  Might just avoid extra_info altogether:

``` python
results.append(cls(channel, row, extra_data[i] if extra_data else None))
```

jtackaberry commented on src/rpc.py:
```diff
@@ -142,37 +142,21 @@ def search(self, channel=None, time=None, cls=Program, **kwargs):
+                time = to_timestamp(time)
+        query_data,extra_data = yield self.channel.rpc('search', channel, time, None, **kwargs)
+        if cls is None:
+            yield  query_data,extra_data
```
Missing space after comma (twice in the previous 3 lines).

jtackaberry commented on src/guide.py:
```diff
@@ -211,20 +201,55 @@ def convert(dt):
+                extra_info = {}
+                for callback in self._program_callbacks:
+                    callback(row['id'], extra_info)
+                extra_data.append(extra_info)
```
Initialize extra_data to None in an `else` instead of unconditionally.  Reads better.

jtackaberry commented on src/guide.py:
```diff
@@ -211,20 +201,55 @@ def convert(dt):
+            for row in query_data:
+                extra_info = {}
+                for callback in self._program_callbacks:
+                    callback(row['id'], extra_info)
```
Why not pass the whole row object?  It would give the callbacks more flexibility as to what they can do.

jtackaberry commented on src/rpc.py:
```diff
@@ -219,6 +203,10 @@ def search(self, channel, time, cls, **kwargs):
         return self.guide.search(channel, time, cls, **kwargs)
 
     @kaa.rpc.expose()
+    def get_grid(self, channels, start_time, end_time, cls):
```
`get_grid` should be removed now.

jtackaberry commented on src/rpc.py:
```diff
@@ -247,3 +235,4 @@ def client_closed(self, client):
         """
         log.info('Client disconnected: %s', client)
         self._clients.remove(client)
+
```
Spurious whitespace.

charrea6 committed:
Remove (un)register_program_callback and replace with a signal.

Couple of small cleans and checks.

jtackaberry commented:
Looks great, Adam.  Thanks!
